File: dsa_leetcode_beginner/0417-pacific-atlantic-water-flow/0417-pacific-atlantic-water-flow.py
class Solution:
    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        # Edge case: if the input is empty, return an empty list
        if not heights:
            return []
        
        # Get the dimensions of the matrix
        m, n = len(heights), len(heights[0])
        
        # Initialize matrices to track cells reachable to the Pacific and Atlantic oceans
        pacificReachable = [[False for _ in range(n)] for _ in range(m)]
        atlanticReachable = [[False for _ in range(n)] for _ in range(m)]
        
        # Define the DFS function that marks reachable cells
        def dfs(row, col, reachable):
            # Mark the current cell as reachable for the current ocean
            reachable[row][col] = True
            
            # Define the possible directions to move (up, down, left, right)
            directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]
            
            for dr, dc in directions:
                # Calculate the new cell's position
                newRow, newCol = row + dr, col + dc
                
                # Check if the new cell is within bounds, unvisited, and its height is >= current cell's height
                if (0 <= newRow < m and 0 <= newCol < n and not reachable[newRow][newCol] and 
                heights[newRow][newCol] >= heights[row][col]):
                    # Recursively perform DFS from the new cell
                    dfs(newRow, newCol, reachable)
        
        # Perform DFS from all cells adjacent to the Pacific ocean
        for i in range(m):
            dfs(i, 0, pacificReachable)
            dfs(i, n - 1, atlanticReachable)
        
        # Perform DFS from all cells adjacent to the Atlantic ocean
        for j in range(n):
            dfs(0, j, pacificReachable)
            dfs(m - 1, j, atlanticReachable)

        print(pacificReachable)
        print(atlanticReachable)
        
        # Collect and return the cells that can reach both oceans
        result = []
        
        for i in range(m):
            for j in range(n):
                if pacificReachable[i][j] and atlanticReachable[i][j]:
                    result.append([i, j])
        
        return result
    

#Question: https://leetcode.com/problems/pacific-atlantic-water-flow
#Blog: https://blog.unwiredlearning.com/pacific-atlantic-water-flow

File: dsa_leetcode_beginner/0417-pacific-atlantic-water-flow/0417-pacific-atlantic-water-flow.java
import java.util.ArrayList;
import java.util.List;

public class Solution {
    public List<List<Integer>> pacificAtlantic(int[][] heights) {
        List<List<Integer>> result = new ArrayList<>();
        if (heights == null || heights.length == 0) {
            return result;
        }

        int m = heights.length;
        int n = heights[0].length;
        boolean[][] pacificReachable = new boolean[m][n];
        boolean[][] atlanticReachable = new boolean[m][n];

        for (int i = 0; i < m; i++) {
            dfs(heights, pacificReachable, i, 0);
            dfs(heights, atlanticReachable, i, n - 1);
        }

        for (int j = 0; j < n; j++) {
            dfs(heights, pacificReachable, 0, j);
            dfs(heights, atlanticReachable, m - 1, j);
        }

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (pacificReachable[i][j] && atlanticReachable[i][j]) {
                    List<Integer> cell = new ArrayList<>();
                    cell.add(i);
                    cell.add(j);
                    result.add(cell);
                }
            }
        }

        return result;
    }

    private void dfs(int[][] heights, boolean[][] reachable, int row, int col) {
        reachable[row][col] = true;
        int[][] directions = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

        for (int[] direction : directions) {
            int newRow = row + direction[0];
            int newCol = col + direction[1];
            if (newRow >= 0 && newRow < heights.length && newCol >= 0 && newCol < heights[0].length &&
                !reachable[newRow][newCol] && heights[newRow][newCol] >= heights[row][col]) {
                dfs(heights, reachable, newRow, newCol);
            }
        }
    }
}
